packages/flutter/lib/src/services/text_input.dart

```diff
@@ -449,4 +537,18 @@ class TextInput {
     );
     return connection;
   }
+
+  static void _ensureInputActionWorksOnPlatform(TextInputAction inputAction) {
+    if (Platform.isIOS) {
+      assert(
+        _iOSSupportedInputActions.contains(inputAction),
+        'The requested TextInputAction "$inputAction" is not supported on iOS.',
+      );
+    } else if (Platform.isAndroid) {
+      assert(
+        _androidSupportedInputActions.contains(inputAction),
+        'The requested TextInputAction "$inputAction" is not supported on Android.',
+      );
+    }
+  }
```

Comment:
If this is debug-only behaviour, then the method should be called "_debugEnsure..." (starting with "debug") and the entire body should be in an assert. And it should be called from an assert. Also this should be very clearly documented; my initial guess based on the docs earlier was that unsupported values would just do the same as "unspecified" on unsupported platforms, the same way that the value that only works on recent iOS versions works. (speaking of which, that should also be documented)

Reply:
@Hixie Can you explain the difference between debug and not debug?  

And you're saying a method with assertions should be called from an assert?  Or you're saying this method should throw exceptions and then the method should be called from an assert?

Do you think the documentation should go at the Widget level or within text_input.dart or both?

Lastly, I think the default that we currently have for an unrecognized action is actually "done", not "unspecified".  Do you think the default for a bad input should be "unspecified"?  Also, do you think that passing in the wrong platforms action should be a silent failure to "unspecified" instead of doing assertions?

This is what I asked about previously with regard to debug logging vs crashing and I went with the judgement call to crash.

Reply:
Flutter has [several modes](https://github.com/flutter/flutter/wiki/Flutter's-modes), of which "debug" and "release" are the prominent ones. In "debug" mode (what you use during development), asserts run. In release mode they do not.

Some methods only do things in debug mode, because in release mode they have no side-effects. The method you wrote here is an example of such a method. We call these debug-only methods. Our style guide [says](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#qualify-variables-and-methods-used-only-for-debugging) that debug-only identifiers should be prefixed with "debug".

The style guide further says that debug methods should only be called from debug code, i.e. from within an assert.

> Do you think the documentation should go at the Widget level or within text_input.dart or both?

Probably both. They're different layers -- someone could write their own framework on top of `text_input.dart`, so we need to document what it does at that level. But similarly, most people will use the widget layer, so the widgets should document what they do too.

>  Lastly, I think the default that we currently have for an unrecognized action is actually "done", not "unspecified". Do you think the default for a bad input should be "unspecified"?

I don't have a strong opinion. I can see the logic in using the platform default, since it's presumably the default for a reason. But also "done" seems to make sense as a default. So... your call.

> Also, do you think that passing in the wrong platforms action should be a silent failure to "unspecified" instead of doing assertions?

I don't have a strong opinion. In release builds the assertions don't do anything, so by using assertions we're saying there's one behaviour in release builds and one in debug builds. So we should think about what we want to have happen in release builds regardless. We should also consider how much we want to support people developing on Android from iOS and vice versa. The gallery has a toggle to switch platforms; how would this work with this feature? Do we expect people to use the platform-specific options? Do we expect them to only use them on platform-specific apps? All these questions are some we should consider here.

Reply:
I renamed the debug method and moved it into an assert.  I added docs to both the TextInputConfiguration and EditableText that explain platform differences for input types. I also updated the engine behavior so that on both iOS and Android if an unrecognized input type value is sent, the OS will go with "default" or "unspecified" depending on which platform you're talking about.